In [1]:
import pandas as pd
import matplotlib.pylab as plt
import numpy as np
import seaborn as sns

### Read 3 Datasets

In [2]:
ds1radiotemp = pd.read_csv("../../dataset_1_radioTemperatures_20210303.csv",sep=';',dtype={'id_audit':np.str})

In [3]:
ds1radiotemp.head()

id_field_values  id_ftp id_audit  value unit        field
0            41310     108      147   23.5    C  TSensor_D4_
1            41311     108      147   23.8    C  TSensor_D5_
2            41312     108      147   26.3    C  TSensor_D6_
3            41313     108      147   23.5    C  TSensor_D8_
4            41314     108      147   23.0    C  TSensor_D9_

In [4]:
ds2powerclass = pd.read_csv("../../dataset_2_powerClass_20210303.csv", sep=";", dtype={'id_audit':np.str})

In [5]:
ds2powerclass.head()

id_trx_status id_audit branch_header       field            value
0           2917       21      Branch A  powerClass  20W [43.00 dBm]
1           3008       21      Branch B  powerClass  20W [43.00 dBm]
2           3099       21      Branch C  powerClass  20W [43.00 dBm]
3           3190       21      Branch D  powerClass  20W [43.00 dBm]
4           3286       21      Branch E  powerClass  40W [46.15 dBm]

In [6]:
ds2powerclass.sort_values(['id_audit','branch_header']).head(10)

id_trx_status id_audit branch_header       field            value
84905      156077628  1000018      Branch A  powerClass  40W [44.80 dBm]
84906      156077727  1000018      Branch B  powerClass  40W [44.80 dBm]
84907      156077826  1000018      Branch C  powerClass  40W [44.80 dBm]
84908      156077925  1000018      Branch D  powerClass  40W [44.80 dBm]
84909      156078041  1000018      Branch E  powerClass  60W [47.80 dBm]
84910      156078140  1000018      Branch F  powerClass  60W [47.80 dBm]
84911      156078240  1000018      Branch G  powerClass  60W [47.80 dBm]
84912      156078351  1000018      Branch H  powerClass  60W [47.80 dBm]
87177      156080912  1000033      Branch A  powerClass  40W [46.00 dBm]
87178      156081014  1000033      Branch B  powerClass  40W [46.00 dBm]

In [7]:
ds3radioid = pd.read_csv("../../dataset_3_radioId_20210303.csv",sep=";", dtype={'id_audit':np.str})

In [8]:
ds3radioid.head()

id_audit  customer  serial
0       21         2   15551
1       23         2   15403
2       25         2   16084
3      147         2   17087
4      746         2   15486

### Check if exists any empty records in radio Temp dataset
* Here we found 641 records without temperatures, among total 3238259
* Sounds mostly due to "Error: Command unsupported"
* TSensor_N22_ is not working at all. Total: 638, Breakdown: 638 
* At this stage, we just remove those records

In [9]:
ds1radiotemp.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3238259 entries, 0 to 3238258
Data columns (total 6 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id_field_values  3238259 non-null  int64  
 1   id_ftp           3238259 non-null  int64  
 2   id_audit         3238259 non-null  object 
 3   value            3237618 non-null  float64
 4   unit             3238259 non-null  object 
 5   field            3238259 non-null  object 
dtypes: float64(1), int64(2), object(3)
memory usage: 148.2+ MB


In [10]:
ds1radiotemp.isnull().sum()

id_field_values      0
id_ftp               0
id_audit             0
value              641
unit                 0
field                0
dtype: int64

In [11]:
ds1radiotemp[ds1radiotemp.value.isnull()].groupby('unit').count()

id_field_values  id_ftp  \
unit                                                                          
Error: Command unsupported                                      596     596   
Error: Link BXP_2048 is not accessible. Please typ                1       1   
Error: Link BXP_2049 is not accessible. Please typ                2       2   
Error: Link BXP_2053 is not accessible. Please typ                2       2   
Error: Link BXP_2054 is not accessible. Please typ                1       1   
Error: Link BXP_2063 is not accessible. Please typ                1       1   
Error: Link BXP_2064 is not accessible. Please typ                1       1   
remote shell died                                                 2       2   
ts                                                                2       2   
unable to hunt BXP_2048/ose_shelld                                9       9   
unable to hunt BXP_2049/ose_shelld                               10      10   
unable to hunt BXP_2050/ose_shelld                                3       3   
unable to hunt BXP_2051/ose_shelld                                2       2   
unable to hunt BXP_2052/ose_shelld                                2       2   
unable to hunt BXP_2053/ose_shelld                                1       1   
unable to hunt BXP_2054/ose_shelld                                1       1   
unable to hunt BXP_2055/ose_shelld                                2       2   
unable to hunt BXP_2058/ose_shelld                                1       1   
unable to hunt BXP_2067/ose_shelld                                1       1   
unsupported                                                       1       1   

                                                    id_audit  value  field  
unit                                                                        
Error: Command unsupported                               596      0    596  
Error: Link BXP_2048 is not accessible. Please typ         1      0      1  
Error: Link BXP_2049 is not accessible. Please typ         2      0      2  
Error: Link BXP_2053 is not accessible. Please typ         2      0      2  
Error: Link BXP_2054 is not accessible. Please typ         1      0      1  
Error: Link BXP_2063 is not accessible. Please typ         1      0      1  
Error: Link BXP_2064 is not accessible. Please typ         1      0      1  
remote shell died                                          2      0      2  
ts                                                         2      0      2  
unable to hunt BXP_2048/ose_shelld                         9      0      9  
unable to hunt BXP_2049/ose_shelld                        10      0     10  
unable to hunt BXP_2050/ose_shelld                         3      0      3  
unable to hunt BXP_2051/ose_shelld                         2      0      2  
unable to hunt BXP_2052/ose_shelld                         2      0      2  
unable to hunt BXP_2053/ose_shelld                         1      0      1  
unable to hunt BXP_2054/ose_shelld                         1      0      1  
unable to hunt BXP_2055/ose_shelld                         2      0      2  
unable to hunt BXP_2058/ose_shelld                         1      0      1  
unable to hunt BXP_2067/ose_shelld                         1      0      1  
unsupported                                                1      0      1

In [12]:
ds1radiotemp[ds1radiotemp.value.isnull()].groupby('field').count()

id_field_values  id_ftp  id_audit  value  unit
field                                                       
TSensor_E20_                1       1         1      0     1
TSensor_I21_                2       2         2      0     2
TSensor_N22_              638     638       638      0   638

In [19]:
ds1radiotemp[ds1radiotemp.field=='TSensor_N22_']['field'].count()

638

In [20]:
ds1radiotemp.dropna(inplace=True)

In [21]:
ds1radiotemp.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3237618 entries, 0 to 3238258
Data columns (total 6 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id_field_values  3237618 non-null  int64  
 1   id_ftp           3237618 non-null  int64  
 2   id_audit         3237618 non-null  object 
 3   value            3237618 non-null  float64
 4   unit             3237618 non-null  object 
 5   field            3237618 non-null  object 
dtypes: float64(1), int64(2), object(3)
memory usage: 172.9+ MB


### Check if exists any empty records in powerclass dataset
* no missing data

In [22]:
ds2powerclass.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366812 entries, 0 to 366811
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id_trx_status  366812 non-null  int64 
 1   id_audit       366812 non-null  object
 2   branch_header  366812 non-null  object
 3   field          366812 non-null  object
 4   value          366812 non-null  object
dtypes: int64(1), object(4)
memory usage: 14.0+ MB


### Check if exists any empty records in radioId dataset
* no missing data

In [23]:
ds3radioid.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74908 entries, 0 to 74907
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id_audit  74908 non-null  object
 1   customer  74908 non-null  int64 
 2   serial    74908 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.7+ MB


### Find out basic data structure
* Total Audit, 67455
    * **each audit contains 48 records** 
    * But data is not full clean. 67455 * 48 = 3237840 (a slight bigger than cleaned data 3237618), less than 3238259 
    * Indicates some audit does not contains 48 records, maybe because of less branches, or because of less sensors
* Total Sensor 71
* Total Serial 27781 
    * *Not sure how Serial is associated to the type of Sensor.* 
    * *Not sure if all sensors of the same type is the same hardware/firmware or different serials incidates different hardware/firmware*
    * *It sounds* **one audit connects one Serials** but ppt said several audit could share the same serials.
    * This field sounds not useable, since it cannot connect to set of sensor clearly, even this might tell the firmware version, and impact temperature
* Total Branch 8
    * *It sounds* **each audit contains 8 branches, therefore, 6 sensors per branch**
* Total PowerClass 366812
    * **each branch contains a powerclass setup** This indicates 6 sensors share the same powerclass setup
    * **Not all audit has powerclass setup** 45854 < 67455 This indicates data is not clean. might remove those dirty audits or use some "default" value
    * **not all branches have powerclass setup (E,F,G,H braches)** This could be because no sensor is installed for E/F/G/H, or not have powerclass setup value. Need further check
    * *Not sure how to link powerclass to a particular set of sensors, among 48 sensors per audit* 
    * We might have to restore powwerclass-sensor relationship from data
* Total 2 customers
    * *Not sure different customers could have different usage/setup, so that temperatures will be highly different* This can be found out by joining the ds1 and ds3.

In [24]:
ds1radiotemp['id_audit'].nunique()

67455

In [25]:
67455 * 48

3237840

In [26]:
ds1radiotemp['field'].nunique()

71

In [27]:
ds3radioid.serial.nunique()

27781

In [32]:
readingPerAuditDf = ds1radiotemp[['id_audit','value']].groupby('id_audit').count().reset_index().sort_values('value')

In [35]:
readingPerAuditDf.describe()

value
count  67455.000000
mean      47.996709
std        0.272347
min       24.000000
25%       48.000000
50%       48.000000
75%       48.000000
max       48.000000

In [37]:
readingPerAuditDf.groupby('value').count()

id_audit
value          
24            7
30            3
48        67445

In [38]:
readingPerAuditDf[readingPerAuditDf.value==24]

id_audit  value
20866  1301295     24
19657  1285950     24
19658  1285951     24
25679  1349385     24
19659  1285952     24
59585   898299     24
51823  1782690     24

In [44]:
ds2powerclass[ds2powerclass.id_audit=='898299']

id_trx_status id_audit branch_header       field            value
30159      126906776   898299      Branch A  powerClass  40W [46.00 dBm]
30160      126906881   898299      Branch B  powerClass  40W [46.00 dBm]
30161      126906985   898299      Branch C  powerClass  40W [46.00 dBm]
30162      126907091   898299      Branch D  powerClass  40W [46.00 dBm]

In [ ]:
ds2powerclass['branch_header'].value_counts()

In [ ]:
ds2powerclass['id_audit'].nunique()

In [ ]:
ds3radioid['id_audit'].nunique()

In [ ]:
ds3radioid.customer.nunique()

### Check temperature distribution
* Some sensors report negative values, this sounds not reasonable. But this might indicate weather temperature could impact record .For example, maybe sensor operates outside, in cold weather. 
* However, we dont have weather temperature. (*same physical location shall have similar weather temperature. If we can get location information from data, it will be good feature for modelling*)
* From quantile, most negative gathers between -1.8 and -9.1, further negative -20 sounds abnormal.
* Data is postive skew. Some extreme postive value exists
* For postive temperatures, quite large range exists, maybe because different type of hardware, sensor, powerclass. need more study

In [ ]:
ds1radiotemp.describe()

In [ ]:
Q1 = ds1radiotemp['value'].quantile(0.25)
Q3 = ds1radiotemp['value'].quantile(0.75)
IQR = Q3 - Q1
Q1out = Q1 - 1.5 * IQR
Q3out = Q3 + 1.5 * IQR
print("25% Quantile: ", Q1, "##75% Quantile: ", Q3, "##IQR: ", IQR, "##Q1 outlier: ", Q1out, "##Q3 outlier:", Q3out)

In [ ]:
ds1radiotemp['value'].skew()

In [ ]:
ds1radiotemp_neg = ds1radiotemp[ds1radiotemp['value'] < 0]

In [ ]:
ds1radiotemp_neg.describe()

In [ ]:
Q1 = ds1radiotemp_neg['value'].quantile(0.25)
Q3 = ds1radiotemp_neg['value'].quantile(0.75)
IQR = Q3 - Q1
Q1out = Q1 - 1.5 * IQR
print("25% Quantile: ", Q1, "##75% Quantile: ", Q3, "##IQR: ", IQR, "##Q1 outlier: ", Q1out)

In [ ]:
#mybins = pd.qcut(ds1radiotemp['value'], q=4, labels = ['lower','low','high','higher'])
cut_bins = [float("-inf"), -30, -20, -10, 0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100,float("inf")]
mybins = pd.cut(ds1radiotemp['value'], bins=cut_bins)

In [ ]:
mybins.value_counts().sort_index()

In [ ]:
ds1radiotemp['bin'] = mybins

In [ ]:
fig, ax = plt.subplots(figsize=(12,8)) 
df2 = ds1radiotemp.groupby('bin').count()
df2['value'].plot(kind='bar', ax=ax)

### Check temperature distribution per audit
* Sounds different audit has large different range. Need found out the reason


In [ ]:
ds1radiotemp.id_audit.unique()

In [ ]:
ds1radiotemp[(ds1radiotemp.id_audit == '147') | (ds1radiotemp.id_audit == '746')
             | (ds1radiotemp.id_audit == '599167')
             | (ds1radiotemp.id_audit == '1970427')
             | (ds1radiotemp.id_audit == '1970491')
             | (ds1radiotemp.id_audit == '1970516')].boxplot(column='value', by='id_audit')

### Check temperature distribution per sensor
* **It indicates 2 differnt temperature segments.** 
* This might explain S1/S2 connets to power amplifier, S3-S9 connects to ohter HW on the radio
* Need find out which sensors are S1/S2, which are S3-S9, or just clustering into 2 groups sounds OK
* When clustering, a little bit concern about outliers might influence, so that sensores are classified into different group.

In [ ]:
pd.set_option('display.max_rows', 500)

In [ ]:
my_orderData = ds1radiotemp.groupby(by=["field"])["value"].median().sort_values(ascending=False)
my_orderData

In [ ]:
ds1radiotemp[['field','value']].groupby('field').median().sort_values(ascending=False, by='value').hist()

In [ ]:
my_order = my_orderData.index

In [ ]:
fig, ax = plt.subplots(figsize=(80,8)) 
ax.set(ylim=(-10, 200))
sns.boxplot(x='field', y='value', data=ds1radiotemp, ax=ax, order=my_order)

In [ ]:
ds1radiotemp.head()

In [ ]:
sortedSensor = ds1radiotemp['field'].value_counts().index

In [ ]:
import matplotlib.ticker as ticker
fig, ax = plt.subplots(figsize=(10,60)) 
ax = plt.axes(xscale='log', yscale='log')
ax.grid();
g=sns.boxplot(x='value', y='field', width=.6, palette="vlag", data=ds1radiotemp, order=sortedSensor)
#sns.stripplot(x="value", y="field", data=ds1radiotemp,
#              size=4, color=".3", linewidth=0)
#g.set_xscale("log")
plt.title("Sensor Record distribution")
#ax.xaxis.set_major_locator(ticker.MultipleLocator(0.5))
#ax.xaxis.set_major_formatter(ticker.ScalarFormatter())
ax.grid()
plt.show()

In [ ]:
fig, ax = plt.subplots(figsize=(80,8)) 
g= sns.countplot(x='field',data=ds1radiotemp, ax=ax, order = ds1radiotemp['field'].value_counts().index)
g.set_yscale("log")
plt.title("Sensor Record Counts")
plt.show()

In [ ]:
sensorCountdf = ds1radiotemp[['field', 'value']].groupby('field').count().reset_index().sort_values(['value'], ascending=False)

In [ ]:
sensorCountdf['logcount'] = np.log10(sensorCountdf.value)

In [ ]:
fig, ax = plt.subplots(figsize=(80,8)) 
sns.barplot(y='logcount',x='field',data=sensorCountdf, ax=ax)

### Check temperature distribution per customer
* 2 customers similar distribution, although the 2nd customer more variance

In [ ]:
ds13join = ds1radiotemp.merge(ds3radioid,sort=False)

In [ ]:
ds13join.head()

In [ ]:
fig, ax = plt.subplots(figsize=(20,8)) 
ax.set(ylim=(-10, 200))
sns.boxplot(x='customer', y='value', data=ds13join, ax=ax)